src/NUnitEngine/nunit.engine/Services/RuntimeFrameworkService.cs

```diff
@@ -109,21 +129,45 @@ private static RuntimeFramework SelectRuntimeFrameworkInner(TestPackage package)
                     throw new NUnitEngineException("Invalid or unknown framework requested: " + frameworkSetting);
 
                 log.Debug($"Requested framework for {package.Name} is {requestedFramework}");
+
+                if (!IsAvailable(frameworkSetting))
+                    throw new NUnitEngineException("Requested framework is not available: " + frameworkSetting);
+
+                package.Settings[EnginePackageSettings.TargetRuntimeFramework] = frameworkSetting;
+
+                return requestedFramework;
             }
-            else
+
+            log.Debug($"No specific framework requested for {package.Name}");
+
+            string imageTargetFrameworkNameSetting = package.GetSetting(InternalEnginePackageSettings.ImageTargetFrameworkName, "");
+            RuntimeType targetRuntime;
+            Version targetVersion;
+
+            if (string.IsNullOrEmpty(imageTargetFrameworkNameSetting))
             {
-                requestedFramework = new RuntimeFramework(RuntimeType.Any, RuntimeFramework.DefaultVersion);
-                log.Debug($"No specific framework requested for {package.Name}");
+                // Assume .NET Framework
+                targetRuntime = RuntimeType.Net;
+                targetVersion = package.GetSetting(InternalEnginePackageSettings.ImageRuntimeVersion, new Version(2, 0));
             }
+            else
+            {
+                FrameworkName frameworkName = new FrameworkName(imageTargetFrameworkNameSetting);
 
-            RuntimeType targetRuntime = requestedFramework.Runtime;
-            Version targetVersion = requestedFramework.FrameworkVersion;
-
-            if (targetRuntime == RuntimeType.Any)
-                targetRuntime = currentFramework.Runtime;
+                switch (frameworkName.Identifier)
+                {
+                    case ".NETFramework":
+                        targetRuntime = RuntimeType.Net;
+                        break;
+                    case ".NETCoreApp":
+                        targetRuntime = RuntimeType.NetCore;
+                        break;
+                    default:
+                        throw new NUnitEngineException("Unsupported Target Framework: " + imageTargetFrameworkNameSetting);
+                }
 
-            if (targetVersion == RuntimeFramework.DefaultVersion)
-                targetVersion = package.GetSetting(InternalEnginePackageSettings.ImageRuntimeVersion, currentFramework.FrameworkVersion);
```

Comment:
Please could you also clarify what the changes in this method are doing too? Sorry - the RuntimeFramework stuff has always been so convoluted, struggling to work out which stage in the process this is affecting here. :-(

Reply:
Sorry. What I've done is take a disparate set of runtime checks, which were being made somewhat later in the process, and centralized them here. For example, we used to get to the point of creating a process and then discover that we either didn't know what runtime to use or that the requested runtime wasn't available. I haven't totally eliminated the late discoveries, but I think I've got 90% of it.

Reply:
Ahh, cool. Thanks!